### PR TITLE
Reduce code size of `thread::set_current`

### PR DESCRIPTION
**Pull request by joboet**
#123265 introduced a rather large binary size regression, because it added an `unwrap()` call on a `Result<(), Thread>`, which in turn pulled its rather heavy `Debug` implementation. This PR fixes this by readding the `rtassert!` that was removed.

### CODE REVIEW

File: library/std/src/thread/mod.rs [hunk 1]
```diff
@@ -684,9 +684,12 @@ thread_local! {
 
 /// Sets the thread handle for the current thread.
 ///
-/// Panics if the handle has been set already or when called from a TLS destructor.
+/// Aborts if the handle has been set already to reduce code size.
 pub(crate) fn set_current(thread: Thread) {
-    CURRENT.with(|current| current.set(thread).unwrap());
+    CURRENT.with(|current| {
+        rtassert!(current.get().is_none());
+        current.set(thread).unwrap();
```
Comment (Nilstrieb) [inline library/std/src/thread/mod.rs:691]: could you `match`+`rtabort` instead to avoid still having an (optimized out) `unwrap`?

**Comment by Nilstrieb** [comment]
I think the push cancelled the try build.
@bors try @rust-timer queue

**Comment by workingjubilee** [comment]
That is ~exactly the number of things buffed as were regressed, and by about as much. Seems to patch the original completely. Nice work!

@bors r+

**Comment by workingjubilee** [comment]
Wait.

@bors r-

@joboet Can you add a comment noting the use of `rtabort!` is load-bearing for about 5kb all on its lonesome?

**Comment by joboet** [comment]
Sure, done!

@bors r=workingjubilee

**Comment by workingjubilee** [comment]
huh?

**Comment by workingjubilee** [comment]
@joboet can you bless the miri tests?

**Comment by joboet** [comment]
I really shouldn't, that looks like actual, albeit completely unrelated UB!

**Comment by joboet** [comment]
I can reproduce the miri issue without this change. I have a working theory, but I'll see if I can minimize the example first.

As it's an unrelated failure, it's not blocking this PR.
@bors retry

File: library/std/src/thread/mod.rs [hunk 2]
```diff
@@ -684,9 +684,14 @@ thread_local! {
...
 ///
-/// Panics if the handle has been set already or when called from a TLS destructor.
+/// Aborts if the handle has been set already to reduce code size.
 pub(crate) fn set_current(thread: Thread) {
-    CURRENT.with(|current| current.set(thread).unwrap());
+    // Using `unwrap` here can add ~3kB to the binary size. We have complete
+    // control over where this is called, so just abort if there is a bug.
+    CURRENT.with(|current| match current.set(thread) {
+        Ok(()) => {}
+        Err(_) => rtabort!("should only be set once"),
```
Comment (RalfJung) [inline library/std/src/thread/mod.rs:693]: This doesn't print the source location I think. So would be good to make the error give a bit of context.
```suggestion
        Err(_) => rtabort!("thread::set_current: should only be set once"),
```

**Comment by joboet** [comment]
@bors r-
If some code running pre-main accesses `thread::current`, this will abort because the thread will be initialized before we get a chance to do so in `init`.

**Comment by RalfJung** [comment]
That just means pre-main code should not do that...?
We control all pre-main code.

**Comment by joboet** [comment]
No, there are ways and means to get around that. We document our limitations [here](https://doc.rust-lang.org/nightly/std/index.html#use-before-and-after-main), `thread::current` before main isn't on the list. And considering how simple it is to support that, I don't see why we shouldn't. I've changed the abort to a noop now.

**Comment by joboet** [comment]
We normally call this in `rt::init` and the thread main function. For threads, nothing can run before it that we don't control, but for main you could use some linker magic. If you then access `thread::current`, it gets initialized to some default value (so that it can be used in foreign threads), meaning that the `set_current` in `rt::init` will abort.

Anyway, I don't feel any particular way about this, as long as we don't `unwrap` on a `Result<(), Thread>`, because that leads to the code size issues this PR aimed to solve.

**Comment by RalfJung** [comment]
> If you then access thread::current, it gets initialized to some default value (so that it can be used in foreign threads), meaning that the set_current in rt::init will abort.

What's the consequence of discarding it the 2nd time? Just a different thread name or anything more serious?

If you change it like this IMO it should be called `try_set_current` or so.

**Comment by workingjubilee** [comment]
:skull: what

**Comment by joboet** [comment]
Sorry for the confusion, I totally derailed the conversation. The PR is now back to the approved state, with an updated abort message as Ralf suggested.
@rustbot ready

**Comment by ChrisDenton** [comment]
> Sorry for the confusion, I totally derailed the conversation. The PR is now back to the approved state, with an updated abort message as Ralf suggested.

lgtm

@bors r+ rollup

**Pull request merged**